src/app/Agent/AgentProbeTemplates.tsx

```diff
@@ -0,0 +1,442 @@
+    );
+    return () => window.clearInterval(id);
+  }, []);
+
+  React.useEffect(() => {
+    const sub = context.target.authFailure().subscribe(() => {
+      setErrorMessage('Auth failure');
+    });
+    return () => sub.unsubscribe();
+  }, [context.target]);
+
+  const displayTemplates = React.useMemo(
+    () => filteredTemplates.map((t: ProbeTemplate) => [t.name, t.xml]),
+    [filteredTemplates]
+  );
+
+  const handleDelete = (rowData) => {
+    addSubscription(
+      context.api
+        .deleteCustomProbeTemplate(rowData[0])
+        .pipe(first())
+        .subscribe(() => {
+          refreshTemplates();
+        })
+    );
+  };
+
+  const handleInsert = (rowData) => {
+    addSubscription(
+      context.api
+        .insertProbes(rowData[0])
+        .pipe(first())
+        .subscribe(() => {})
+    );
+  };
+
+  const actionResolver = (rowData: IRowData, extraData: IExtraData) => {
+    if (typeof extraData.rowIndex == 'undefined') {
+      return [];
+    }
+    let actions = [
+      {
+        title: 'Insert Probes...',
+        onClick: (event, rowId, rowData) => handleInsert(rowData),
+      },
+    ] as IAction[];
+    actions = actions.concat([
+      {
+        isSeparator: true,
+      },
+      {
+        title: 'Delete',
+        onClick: (event, rowId, rowData) => handleDeleteButton(rowData),
+      },
+    ]);
+    return actions;
+  };
+
+  const handleModalToggle = () => {
+    setModalOpen((v) => {
+      if (v) {
+        setUploadFile(undefined);
+        setUploadFilename('');
+        setUploading(false);
+      }
+      return !v;
+    });
+  };
+
+  const handleFileChange = (value, filename) => {
+    setFileRejected(false);
+    setUploadFile(value);
+    setUploadFilename(filename);
+  };
+
+  const handleUploadSubmit = () => {
+    if (!uploadFile) {
+      window.console.error('Attempted to submit template upload without a file selected');
+      return;
+    }
+    setUploading(true);
+    addSubscription(
+      context.api
+        .addCustomProbeTemplate(uploadFile)
+        .pipe(first())
+        .subscribe((success) => {
+          setUploading(false);
+          if (success) {
+            setUploadFile(undefined);
+            setUploadFilename('');
+            setModalOpen(false);
+            refreshTemplates();
+          }
+        })
+    );
+  };
+
+  React.useEffect(() => {
+    addSubscription(
+      context.notificationChannel
+        .messages(NotificationCategory.ProbeTemplateUploaded)
+        .subscribe((v) => refreshTemplates())
+    );
+  }, [addSubscription, context, context.notificationChannel, setTemplates]);
+
+  React.useEffect(() => {
+    addSubscription(
+      context.notificationChannel.messages(NotificationCategory.TemplateDeleted).subscribe((v) => refreshTemplates())
+    );
+  }, [addSubscription, context, context.notificationChannel, setTemplates]);
+
+  const handleUploadCancel = () => {
```

Comment:
I think, from past experience, many of these functions can be React.useCallback functions like 
```
const handleUploadCancel = React.useCallback() => {
  setUploadFile(undefined);
  setUploadFilename('');
  setModalOpen(false);
}, [setUploadFile, setUploadFilename, setModalOpen])
```
with the correct dependencies array and such.